examples/browser/test/run-test.js

```diff
@@ -0,0 +1,114 @@
+// @ts-check
+
+/**
+ * @typedef {Object} TestOptions
+ * @property {import('puppeteer').LaunchOptions} launch
+ * @property {Partial<Parameters<import('mocha/lib/cli/collect-files')>[0]>} collectFiles
+ */
+
+/**
+ * @param {Partial<TestOptions>} [options]
+ */
+module.exports = async options => {
+    const launch = options && options.launch;
+    const collectFiles = {
+        ignore: options && options.collectFiles && options.collectFiles.ignore || [],
+        extension: options && options.collectFiles && options.collectFiles.extension || [],
+        file: options && options.collectFiles && options.collectFiles.file || [],
+        recursive: options && options.collectFiles && options.collectFiles.recursive || false,
+        sort: options && options.collectFiles && options.collectFiles.sort || false,
+        spec: options && options.collectFiles && options.collectFiles.spec || []
+    };
+
+    // quick check whether test files exist
+    require('mocha/lib/cli/collect-files')(collectFiles);
+    const puppeteer = require('puppeteer');
+
+    const address = await require('../src-gen/backend/main');
+    const browser = await puppeteer.launch(launch);
+
+    const page = await browser.newPage();
+    page.on('dialog', dialog => dialog.dismiss());
+    page.on('pageerror', console.error);
+
+    let theiaLoaded = false;
+    page.exposeFunction('fireDidUnloadTheia', () => theiaLoaded = false);
+    const preLoad = () => {
+        if (theiaLoaded) {
+            return;
+        }
+        console.log('loading chai...');
+        theiaLoaded = true;
+        page.addScriptTag({ path: require.resolve('chai/chai.js') });
+        page.evaluate(() =>
+            window.addEventListener('beforeunload', () => window['fireDidUnloadTheia']())
+        );
+    }
+    page.on('frameattached', preLoad);
+    page.on('framenavigated', preLoad);
+
+    page.on('load', async () => {
+        console.log('loading mocha...');
+        // replace console.log by theia logger for mocha
+        await page.waitForFunction(() => !!window['theia']['@theia/core/lib/common/logger'].logger, {
+            timeout: 30 * 1000
+        });
+        await page.addScriptTag({ path: require.resolve('mocha/mocha.js') });
```

Comment:
@amiramw I went for running mocha tests inside the browser, since it makes debugging loop easier, but you can define your own puppeteer runner with mocha in Node.js if required